test/TestApp30.Tests/BasicTestWithCorrelation.cs

```diff
@@ -0,0 +1,318 @@
+
+            Assert.Equal("40d1a5a08a68c0998e4a3b7c91915ca6", req.Context.Operation.Id);
+            Assert.Equal("40d1a5a08a68c0998e4a3b7c91915ca6", exc.Context.Operation.Id);
+            
+            Assert.Equal(req.Id, exc.Context.Operation.ParentId);
+            Assert.Equal("|40d1a5a08a68c0998e4a3b7c91915ca6.b9e41c35_1.", req.Context.Operation.ParentId);
+            Assert.Contains("|40d1a5a08a68c0998e4a3b7c91915ca6.", req.Id);
+
+            Assert.Equal("http://localhost/" + url, req.Url.ToString());
+            Assert.False(req.Success);
+        }
+
+        private HttpRequestMessage CreateRequestMessage(Dictionary<string, string> requestHeaders)
+        {
+            HttpRequestMessage httpRequestMessage = new HttpRequestMessage();
+            httpRequestMessage.Method = HttpMethod.Get;
+            if (requestHeaders != null)
+            {
+                foreach (var h in requestHeaders)
+                {
+                    httpRequestMessage.Headers.Add(h.Key, h.Value);
+                }
+            }
+
+            return httpRequestMessage;
+        }
+
+        [Fact]
+        public async Task ARequestSuccessWithNonW3CCompatibleRequestId()
+        {
+            // Arrange
+            var client = _factory.CreateClient();
+            var url = "Home/Index";
+
+            // Act
+            Dictionary<string, string> requestHeaders = new Dictionary<string, string>()
+                {
+                    { "Request-Id", "|noncompatible.b9e41c35_1."}
+                };
+            var request = CreateRequestMessage(requestHeaders);
+            request.RequestUri = new Uri(client.BaseAddress + url);
+            
+            var response = await client.SendAsync(request);
+
+            // Assert
+            response.EnsureSuccessStatusCode();
+            this.output.WriteLine(response.Content.ReadAsStringAsync().Result);
+
+            Task.Delay(1000).Wait();
+            var items = _factory.sentItems;
+            PrintItems(items);
+            // 1 Trace from Ilogger, 1 Request
+            Assert.Equal(2, items.Count);
+
+            var req = GetFirstTelemetryOfType<RequestTelemetry>(items);
+            var trace = GetFirstTelemetryOfType<TraceTelemetry>(items);
+            Assert.NotNull(req);
+            Assert.NotNull(trace);
+
+            Assert.NotEqual("noncompatible", req.Context.Operation.Id);
+            Assert.NotEqual("noncompatible", trace.Context.Operation.Id);
+
+            Assert.Equal("|noncompatible.b9e41c35_1.", req.Context.Operation.ParentId);
+            Assert.Contains($"|{req.Context.Operation.Id}.", req.Id);
+            Assert.Equal(req.Id, trace.Context.Operation.ParentId);
+            Assert.Equal("noncompatible", req.Properties["ai_legacyRootId"]);
+
+            Assert.Equal("http://localhost/" + url, req.Url.ToString());
+            Assert.True(req.Success);
+        }
+
+        private T GetFirstTelemetryOfType<T>(ConcurrentBag<ITelemetry> items)
```

Comment:
if it's a bag, you should not really ask for first item as order is not guaranteed. If order mettars this probably should be a queue?

Reply:
order dont matter for tests. i refactored to get all items, and validate that only expected count is found.